### ISSUE TITLE
CMake whitespace error CMP0004 with CentOS 7 rpmbuild system LDFLAGS

### ISSUE THREAD

**Issue by antcodd** [actor-framework/actor-framework#718]
CAF does not build using the LDFLAGS supplied by rpmbuild in CentOS 7 because of trailing whitespace and CMake policy CMP0004.

This is similar to #653, but instead with externally added flags. CMakelists.txt has also changed significantly since that issue.

After some investigation this is because of the way RPM (4.11.3) itself is constructing the LDFLAGS supplied with ``%configure``:
```
$ rpm --showrc | grep relro
-14: __global_ldflags   -Wl,-z,relro %{_hardened_ldflags}
```
where ``%{_hardened_ldflags}`` is not set.

For context I was attempting to build a local RPM of Bro git master which currently builds libcaf master as a 3rdparty submodule. 

I have since discovered the slightly outdated OBS CAF repository with spec file. That doesn't use %configure so the default system LDFLAGS aren't set, presumably because of a lack of supported arguments. A quick hack forward-port seems to build on CentOS 7.4 now (GCC 4.8.5), except for some packaged file issues. As a separate issue it might make sense to update that and pull in parts of the %configure macro. Bro requires caf 0.15.5 so I can't use the prebuilt versions.

```
********** Begin CAF External Project CMake Output ************

-- The C compiler identification is GNU 4.8.5
-- The CXX compiler identification is GNU 4.8.5
-- Check for working C compiler: /bin/cc
-- Check for working C compiler: /bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /bin/c++
-- Check for working CXX compiler: /bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Found g++ version 4.8
-- Found OpenSSL: /usr/lib64/libssl.so;/usr/lib64/libcrypto.so (found version "1.0.2k") 
-- Performing Test OPENSSL_CORRECT_VERSION_NUMBER
-- Performing Test OPENSSL_CORRECT_VERSION_NUMBER - Success
-- Found Doxygen: /bin/doxygen (found version "1.8.5") 
-- 
====================|  Build Summary  |====================

CAF version:           0.15.7

Build type:            RelWithDebInfo
Build static:          no
Build static only:     no
Build static runtime:  no
Runtime checks:        no
Log level:             none
With mem. mgmt.:       yes
With exceptions:       yes

Build I/O module:      yes
Build tools:           no
Build examples:        no
Build unit tests:      no
Build benchmarks:      no
Build OpenCL:          no
Build OpenSSL:         yes
Build Python:          no

CXX:                   /bin/c++
CXXFLAGS:              -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic  -O2 -g -DNDEBUG
LDFLAGS:               -Wl,-z,relro 

Source directory:      /home/eda/rpmbuild/BUILD/bro-2.5-719/aux/broker/3rdparty/caf
Build directory:       /home/eda/rpmbuild/BUILD/bro-2.5-719/build/aux/broker/caf-build
Executable path:       
Library path:          
Install prefix:        /usr
Generator:             Unix Makefiles

===========================================================

-- Configuring done
CMake Error at libcaf_io/CMakeLists.txt:44 (add_library):
  Target "libcaf_io_shared" links to item "-Wl,-z,relro " which has leading
  or trailing whitespace.  This is now an error according to policy CMP0004.


CMake Error at libcaf_io/CMakeLists.txt:44 (add_library):
  Target "libcaf_io_shared" links to item "-Wl,-z,relro " which has leading
  or trailing whitespace.  This is now an error according to policy CMP0004.


CMake Warning (dev) in libcaf_io/CMakeLists.txt:
  Policy CMP0022 is not set: INTERFACE_LINK_LIBRARIES defines the link
  interface.  Run "cmake --help-policy CMP0022" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  Target "libcaf_io_shared" has an INTERFACE_LINK_LIBRARIES property.  This
  should be preferred as the source of the link interface for this library
  but because CMP0022 is not set CMake is ignoring the property and using the
  link implementation as the link interface instead.

  INTERFACE_LINK_LIBRARIES:

    -Wl,-z,relro ;libcaf_core_shared

  Link implementation:

    -Wl,-z,relro;libcaf_core_shared

This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at libcaf_io/CMakeLists.txt:44 (add_library):
  Target "libcaf_io_shared" links to item "-Wl,-z,relro " which has leading
  or trailing whitespace.  This is now an error according to policy CMP0004.


CMake Error at libcaf_io/CMakeLists.txt:44 (add_library):
  Target "libcaf_io_shared" links to item "-Wl,-z,relro " which has leading
  or trailing whitespace.  This is now an error according to policy CMP0004.


-- Generating done
-- Build files have been written to: /home/eda/rpmbuild/BUILD/bro-2.5-719/build/aux/broker/caf-build

*********** End CAF External Project CMake Output *************


CMake Error at aux/broker/CMakeLists.txt:160 (message):
  CAF CMake configuration failed
```
Note the trailing whitespace after ``LDFLAGS:               -Wl,-z,relro ``. Attempting to force CMP0004 OLD is deprecated and difficult because it seems to need to be done per subdirectory (there are many more such errors in other CAF subdirectories if CMake is given the chance to get to them). The INTERFACE_LINK_LIBRARIES dev warning appears to be caused by ``target_link_libraries`` itself and goes away when the whitespace issue is fixed.

**Steps to reproduce outside with a regular CAF git checkout**
1. ``mkdir build && cd build``
2. ``LDFLAGS="-Wl,-z,relro " cmake ..``

**Possible solutions**
It looks to me like loading the LDFLAGS environment variable may be unnecessary. In later versions of the CMake manual LDFLAGS is documented as populating the default LINKER_FLAGS values. In CMake 2.6.8.12 this also seems to be the case but is undocumented. Even with the trailing whitespace here it works fine. I have very limited experience with CMake though.

```
$ git diff CMakeLists.txt
diff --git a/CMakeLists.txt b/CMakeLists.txt
index b11f1ae..ab97250 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -132,7 +132,7 @@ endif()
 #                        respect environment variables                         #
 ################################################################################

-set(LDFLAGS $ENV{LDFLAGS})
+set(LDFLAGS "")

 ################################################################################
 #                              get version of CAF                              #
```
```
$ mkdir build/ && cd build/
$ LDFLAGS="-Wl,-z,relro " cmake ..
$ cat build/CMakeCache.txt | grep relro
CMAKE_EXE_LINKER_FLAGS:STRING=' -Wl,-z,relro '
CMAKE_MODULE_LINKER_FLAGS:STRING=' -Wl,-z,relro '
CMAKE_SHARED_LINKER_FLAGS:STRING=' -Wl,-z,relro '
```
The flags also seem to appear in the linker command line.

Another option is something like this, but I'm not sure if it covers all cases.
```
string(STRIP "$ENV{LDFALGS}" LDFLAGS_STRIPWS)
set(LDFLAGS ${LDFLAGS_STRIPWS})
```

**Comment by antcodd** [comment]
This seems to work with rpmbuild (with the commit applied as a patch), though I don't think all the extra flags added in CMakeLists.txt are getting passed through to the subdirectories due to the variable rename.

I now have an issue with the system supplied CXXFLAGS and the lack of ``-std=c++11``, because CMakeLists.txt doesn't override it if it is already set.
I'm not sure if a similar automatic use of CXXFLAGS applies. My understanding is LDFLAGS and CXXFLAGS populate the default values on the initial build but can be overridden. Perhaps there is a way of auto-populating the values while still allowing the user to override them in the CMake cache if they don't suit? e.g. can you determine if this is the initial build?

Bro does something like this, but I'm not sure what happens if you reconfigure the build multiple times:
```
# Append our extra flags to the existing value of CXXFLAGS.
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${EXTRA_FLAGS}")
```

I can open a separate issue for CXXFLAGS if you like.

**Comment by Neverlord** [comment]
Should be fixed via #721.
